test/test_modules.py

```diff
@@ -205,6 +206,120 @@ def test_check_inplace(self, device, dtype, module_info):
             output_ip.backward(grad)
             self.assertEqual(input_args[0].grad, input_arg_copy[0].grad)
 
+    def _retain_grad(self, obj):
+        # gradients needs to be retained to check for grad. This is useful when
+        # non-leafs are present in the graph.
+        if isinstance(obj, dict):
+            for i in obj.values():
+                self._retain_grad(i)
+        elif isinstance(obj, (tuple, list)):
+            for i in obj:
+                self._retain_grad(i)
+        elif isinstance(obj, torch.Tensor) and obj.requires_grad:
+            obj.retain_grad()
+
+    def _get_grads(self, obj):
+        if isinstance(obj, (tuple, list)):
+            return tuple(self._get_grads(o) for o in obj)
+        elif isinstance(obj, dict):
+            return {name: self._get_grads(o) for name, o in obj.items()}
+        elif isinstance(obj, torch.Tensor) and obj.requires_grad:
+            return obj.grad
+
+    def _zero_grad(self, obj):
+        if isinstance(obj, (tuple, list)) or isgenerator(obj):
+            for o in obj:
+                self._zero_grad(o)
+        elif isinstance(obj, dict):
+            for o in obj.values():
+                self._zero_grad(o)
+        elif isinstance(obj, (torch.Tensor, torch.nn.Parameter)) and obj.grad is not None:
+            obj.grad.detach_()
+            obj.grad.zero_()
```

Comment:
Hm I'm starting to see some patterns here - I wonder if there's some more generic way we can do arbitrary things over all inputs?